src/threadpool.c

```diff
@@ -388,6 +477,14 @@ int uv_cancel(uv_req_t* req) {
   case UV_WORK:
     loop =  ((uv_work_t*) req)->loop;
     wreq = &((uv_work_t*) req)->work_req;
+    /* This is only the case when work is queued without an after_work_cb. In
+     * this case, safe removal cannot be guaranteed.  Returning EINVAL is just
```

Comment:
W.r.t. "safe removal", how is that? If it's still sitting in the work queue, it's still cancelable, right?

Reply:
Here's an example:
```c
void work_cb(uv_work_t* req) {
  free(req);
}

/* From another thread */
void queue_work() {
  uv_work_t* req = malloc(sizeof(*req));
  uv_queue_work(NULL, req, work_cb, NULL);
  /* Do something */
  uv_cancel((uv_req_t*) req);
}
```
In the above there is a chance that `work_cb()` runs before `uv_cancel()` can remove the `req` from the queue. So it's not that we can't remove it from the queue, it's that there's a chance we access invalid memory and I can't think of a way to enforce safe access.

Basically, the user would need to guarantee that `req` will not be free'd before any call to `uv_cancel()` completes, but I feel like that's just a minefield. I'll admit that's a bit paranoid and maybe we shouldn't limit users that way. If you think we can trust the user I'll be happy to make it so req's can be cancelled.

Reply:
Forgot to add that if we allow a call to `uv_cancel()` on a `uv_work_t` that's only been queued with a `uv_work_cb`, then the call to `uv_cancel()` is implicitly thread-safe. So we'd need to clearly document that specific case.

Hypothetically the user should be able to `uv_cancel()` any `uv_work_t`, but I'd want to write some tests to make sure there would be no unexpected side effects.